### PR TITLE
Handle dict of metrics in distributed training

### PR DESCRIPTION
**Pull request by froody**
When training with a model with a subclass `tf.keras.metrics.Metric` which which returns a dict, e.g.

```
class CustomMetric(tf.keras.metrics.Metric):
    def result(self):
        return {"good", tf.convert_to_tensor([1.0]), "bad": tf.convert_to_tensor([0.0])}
```

I got an exception in `merge_fn_wrapper` when trying to call `array_ops.identity` on a `Dict[str, tf.Tensor]`. This fix allows me to do distributed training with the above metric.

### CODE REVIEW

**Comment by sr5434** [comment]
This isn't my PR
